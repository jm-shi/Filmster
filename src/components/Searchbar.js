import React from 'react';
import { connect } from 'react-redux';
import { fetchMovie, fetchMovieDetails } from '../actions/movieActions';
import { history } from '../AppRouter';

class Searchbar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      query: ''
    };
  }
  onQueryChange = e => {
    const query = e.target.value;
    this.setState(() => ({ query }));
  };
  onSubmit = e => {
    e.preventDefault();

    const movie = this.state.query;
    if (movie) {
      Promise.resolve(this.props.onFetchMovie(movie)).then(() => {
        console.log('searchbar.js: this.props gives us  ', this.props);
        this.props.onFetchMovieDetails(this.props.movie);
        history.push('/details');
      });
    }
  };
  render() {
    return (
      <form className="searchbar" onSubmit={this.onSubmit}>
        <h1>Search for movies and TV shows</h1>
        <input value={this.state.query} onChange={this.onQueryChange} />

        <button>Search</button>
      </form>
    );
  }
}

const mapStateToProps = state => {
  const { movie } = state.movieReducer;
  return {
    movie
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onFetchMovie: movie => dispatch(fetchMovie(movie)),
    onFetchMovieDetails: movieID => dispatch(fetchMovieDetails(movieID))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Searchbar);
