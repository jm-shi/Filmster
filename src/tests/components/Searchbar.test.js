import React from 'react';
import { shallow } from 'enzyme';
import Searchbar from '../../components/Searchbar';

let mockHistory, wrapper;
const event = {
  preventDefault() {},
  target: { value: 'Movie title' }
};

beforeEach(() => {
  mockHistory = { push: jest.fn() };
  wrapper = shallow(<Searchbar history={mockHistory} />);
});

test('should render Searchbar', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should update Searchbar state upon query change', () => {
  let queryState = wrapper.state('query');
  expect(queryState).toEqual('');
  wrapper.find('input').prop('onChange')(event);
  queryState = wrapper.state('query');
  expect(queryState).toEqual('Movie title');
});

test('should call history.push upon submit and movie title exists', () => {
  wrapper.setState({ query: 'Movie title' });
  wrapper.find('form').simulate('submit', event);
  expect(mockHistory.push).toHaveBeenCalledTimes(1);
});

test('should not call history.push upon submit and movie title is blank', () => {
  wrapper.find('form').simulate('submit', event);
  expect(mockHistory.push).toHaveBeenCalledTimes(0);
});
