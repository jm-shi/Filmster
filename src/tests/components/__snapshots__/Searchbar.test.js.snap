// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Searchbar 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Searchbar
    history={
        Object {
            "push": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "",
        <div
          className="searchbar-container"
>
          <input
                    className={undefined}
                    onChange={[Function]}
                    placeholder="Enter a movie title"
                    value=""
          />
          <button
                    className="button"
          >
                    Search
          </button>
</div>,
      ],
      "className": undefined,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className={undefined}
              onChange={[Function]}
              placeholder="Enter a movie title"
              value=""
/>,
            <button
              className="button"
>
              Search
</button>,
          ],
          "className": "searchbar-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "onChange": [Function],
              "placeholder": "Enter a movie title",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
              "className": "button",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "",
          <div
            className="searchbar-container"
>
            <input
                        className={undefined}
                        onChange={[Function]}
                        placeholder="Enter a movie title"
                        value=""
            />
            <button
                        className="button"
            >
                        Search
            </button>
</div>,
        ],
        "className": undefined,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className={undefined}
                onChange={[Function]}
                placeholder="Enter a movie title"
                value=""
/>,
              <button
                className="button"
>
                Search
</button>,
            ],
            "className": "searchbar-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": undefined,
                "onChange": [Function],
                "placeholder": "Enter a movie title",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
                "className": "button",
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
